src/core/public/plugins/plugin.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { DiscoveredPlugin, PluginName } from '../../server';
+import { PluginInitializerContext, PluginSetupContext } from './plugin_context';
+import { loadPluginBundle } from './plugin_loader';
+
+/**
+ * The `plugin` export at the root of a plugin's `public` directory should conform
+ * to this interface.
+ */
+export type PluginInitializer<TSetup, TDependencies extends Record<string, unknown>> = (
+  core: PluginInitializerContext
+) => {
+  setup: (core: PluginSetupContext, dependencies: TDependencies) => TSetup | Promise<TSetup>;
+  stop?: () => void;
+};
+
+/**
+ * Lightweight wrapper around discovered plugin that is responsible for instantiating
+ * plugin and dispatching proper context and dependencies into plugin's lifecycle hooks.
+ * @internal
+ */
+export class Plugin<
+  TSetup = unknown,
+  TDependenciesSetup extends Record<PluginName, unknown> = Record<PluginName, unknown>
+> {
+  public readonly name: DiscoveredPlugin['id'];
+  public readonly configPath: DiscoveredPlugin['configPath'];
+  public readonly requiredDependencies: DiscoveredPlugin['requiredPlugins'];
+  public readonly optionalDependencies: DiscoveredPlugin['optionalPlugins'];
+  private initializer?: PluginInitializer<TSetup, TDependenciesSetup>;
+  private instance?: ReturnType<PluginInitializer<TSetup, TDependenciesSetup>>;
+
+  constructor(
+    readonly discoveredPlugin: DiscoveredPlugin,
+    private readonly initializerContext: PluginInitializerContext
+  ) {
+    this.name = discoveredPlugin.id;
+    this.configPath = discoveredPlugin.configPath;
+    this.requiredDependencies = discoveredPlugin.requiredPlugins;
```

Comment:
Maybe we should just adopt the term/convention `plugins` instead of `dependencies`. I know I've been inconsistent with it, but if we're going to ever be consistent with it now-ish is the best time. We can address that in a separate PR though.

Reply:
Sounds good. I'll do this in an immediate follow-up so that the changes to the server side are separate from this PR.